### PR TITLE
Issue %patchN deprecation warning just once

### PR DESCRIPTION
**Pull request by dmnks**
Avoid polluting the build log with this warning repeated as many times as there are %patchN lines.  Once per build should be enough for the packager to notice (there's also a warning summary now).

Basically, make the warning context-independent, meaning that it doesn't matter which line caused it (so don't even print it), we just want to warn about such usage being present in the SPEC.

Suppressing duplicate log messages could be a useful feature on its own so we may eventually want to implement this in rpmlog() with something like a RPMLOG_NODUP flag (and perhaps print the number of suppressed messages in the summary, too) but until then, just keep it simple and use a static int.

Rather than adding another %patch to hello2-suid.spec which is also used elsewhere in the test suite, just make a simplified clone dedicated to %patch testing.

Fixes: #2383

### CODE REVIEW

**Comment by dmnks** [comment]
This is how the alternative rpmlog-native solution might look like: https://github.com/dmnks/rpm/commit/37a28123977570f4eb0a2162e2d73bd9529f95dd

**Review by kaboussi — APPROVED** [review]

**Comment by pmatilai** [comment]
Static variables are best avoided, for one they wont work correctly for the case where you query/build multiple specs on a single run.

A simple alternative would be adding a special purpose counter to the spec struct, and log a message based on that towards the end of parsePrep() where you can then also log the number of times emitted.

rpmlog() based deduplicator could be handy though, there are some pathological cases elsewhere where we end up flooding other (more) relevant messages due to stupidly repeated errors. But for that, I'd look at only suppressing consecutive identical messages, which has multiple benefits:
- you only ever need to compare the current and previous message
- identical messages over a larger timespan and/or with interleaving other messages are probably something you'd want to see anyhow

Basically, whenever a message is logged, compare to the previous one and if identical, bump a counter. When a non-identical message arrives, log a "previous warning/error suppressed N times", reset counter and proceed with the new message as usual. Bonus points for an additional time-stamp and using that for additional heuristics: if a log message is repeated within N seconds, you probably want to suppress but otherwise you may want to show it regardless.

**Comment by dmnks** [comment]
> Static variables are best avoided, for one they wont work correctly for the case where you query/build multiple specs on a single run.

Argh... The multi-spec case somehow didn't occur to me... yet it's pretty obvious :sweat_smile:

> A simple alternative would be adding a special purpose counter to the spec struct, and log a message based on that towards the end of parsePrep() where you can then also log the number of times emitted.

Yup, or even simpler: just add an int pointer argument to `doPatchMacro()` where `parsePrep()` would pass a counter which the function would increment, and then do whatever (print the message, count, etc.) at the end.

Both of these solutions have the downside of "leaking" this particular warning outside of `doPatchMacro()`, though. It might just be fine (and better than `static`) as a temporary solution.

> Basically, whenever a message is logged, compare to the previous one and if identical, bump a counter. When a non-identical message arrives, log a "previous warning/error suppressed N times", reset counter and proceed with the new message as usual. This will need some sort of flushing though to ensure the "repeated" message gets shown (but at least for build, I think we'll get it for free from the "build warnings" log message) . Bonus points for an additional time-stamp and using that for additional heuristics: if a log message is repeated within N seconds, you probably want to suppress but otherwise you may want to show it regardless.

Nice idea! Thanks for the detailed description, I couldn't have wished for more guidance, really :smile:

**Comment by dmnks** [comment]
OK, reworked based on:

> A simple alternative would be adding a special purpose counter to the spec struct, and log a message based on that towards the end of parsePrep() where you can then also log the number of times emitted.

This simple solution will unblock the 4.18.1 release. Later, I'll do this in `rpmlog.c` as suggested in the same comment (in fact, I have a half-functioning WIP already).

File: build/parsePrep.c [hunk 1]
```diff
@@ -450,6 +448,20 @@ int parsePrep(rpmSpec spec)
...
+    if (num) {
+	const char *msg =
+	    _("%%patchN is deprecated, use %%patch N (or %%patch -P N)\n");
+	rpmlog(RPMLOG_WARNING, msg);
+	/* Issue one additional warning for any remaining conversions */
+	if (num == 2)
+	    rpmlog(RPMLOG_WARNING, msg);
+	if (num > 2)
+	    rpmlog(RPMLOG_WARNING,
+		   _("--- Last warning suppressed %i times ---\n"), num - 1);
```
Comment (pmatilai) [inline build/parsePrep.c:462]: Oh, I would've just condensed it into one "patchN deprecated (repeated X times)" message. This is pretty fancy handling for just one message :smile:

Comment (dmnks) [reply build/parsePrep.c:462]: OK, true :smile: This is basically the format of the general rpmlog-native solution that I've started separately, but for now (or for this particular message), it's indeed too elaborate. I'll push a simplified version later today/tomorrow. Thanks!

**Comment by dmnks** [comment]
How about this one?

**Comment by pmatilai** [comment]
Seems as simple as it's going to get :laughing:
